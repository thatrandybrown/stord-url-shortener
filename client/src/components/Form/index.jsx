import { useState } from "react";
import Error from "../Error";
import Interstitial from "../Interstitial";
import { postShortLink } from "../../controllers/shortLink";

const state = {
  Loading: <Interstitial />,
  PostError: (
    <Error
      {...{
        // link: false,
        // heading: null,
        detail: "There was an error submitting your request. Please try again.",
      }}
    />
  ),
};

const Form = ({ setData }) => {
  const [url, setUrl] = useState("");
  const [display, setDisplay] = useState("");

  const post = (event) => {
    event.preventDefault();
    setDisplay("Loading");
    postShortLink(url)
      .then((rsp) => (rsp.ok ? rsp : Promise.reject("service call failed")))
      .then((rsp) => {
        setDisplay("");
        setUrl("");
        return rsp.json();
      })
      .then(setData)
      .catch((err) => setDisplay("PostError"));
  };

  return display === "PostError" ? (
    state[display]
  ) : (
    <>
      <p>
        Shorten your link: provide a full url and get a shortened link for
        sharing.
      </p>
      <form onSubmit={post}>
        <label htmlFor="url">URL:</label>
        <input
          id="url"
          type="url"
          name="url"
          required
          onChange={(event) => setUrl(event.target.value)}
          value={url}
        />
        {/* <label htmlFor="submit">
          Submit:
        </label> */}
        <input
          type="submit"
          name="submit"
          value="Shorten"
          disabled={url === ""}
          aria-label="Shorten"
        />
      </form>
      {display === "Loading" && state[display]}
    </>
  );
};

export default Form;
