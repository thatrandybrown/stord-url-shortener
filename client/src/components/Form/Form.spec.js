import { render, fireEvent, act, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Form from "./";

jest.mock("../../controllers/shortLink", () => {
  return {
    postShortLink: (url) => {
      // console.log("called postShortLink with " + url);
      if (url === "https://www.google.com")
        return Promise.resolve({
          ok: true,
          json: () => ({
            shortCode: "abcdef",
            target: "https://www.google.com",
          }),
        });
      else if (url === "https://developer.mozilla.org/en-US/")
        return Promise.resolve({
          ok: false,
          status: 500,
          json: () => ({ message: "Internal Service Error" }),
        });
      else if (url === "https://reactjs.org/")
        return Promise.reject("Something went wrong!");
      else return Promise.resolve({ status: 404, message: "Page Not Found" });
    },
  };
});

test("renders instructions", () => {
  const { container, getByText } = render(<Form />);
  const instructions = getByText(
    /Shorten your link: provide a full url and get a shortened link for sharing./i
  );
  expect(instructions).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test("renders the form correctly", () => {
  const { getByText, getByLabelText, getByRole } = render(<Form />);
  const UrlLabel = getByText(/URL:/i);
  expect(UrlLabel).toBeInTheDocument();
  const urlField = getByLabelText(/URL:/i);
  expect(urlField).toHaveAttribute("type", "url");
  const submitButton = getByRole("button", { name: "Shorten" });
  expect(submitButton).toBeInTheDocument();
});

test("disables form submission before url is provided", () => {
  const { getByLabelText, getByRole } = render(<Form />);
  const urlField = getByLabelText(/URL:/i);
  expect(urlField).toHaveAttribute("required");
  fireEvent.change(urlField, { target: { value: "" } });
  const submitButton = getByRole("button", { name: "Shorten" });
  expect(submitButton).toHaveAttribute("disabled");
});

test("display form input as the input value", () => {
  const { getByLabelText } = render(<Form />);
  const urlField = getByLabelText(/URL:/i);
  fireEvent.change(urlField, { target: { value: "https://www.google.com" } });
  expect(urlField.value).toBe("https://www.google.com");
});

test("enables form submission when url is provided", () => {
  const { getByLabelText, getByRole } = render(<Form />);
  const urlField = getByLabelText(/URL:/i);
  fireEvent.change(urlField, { target: { value: "abcd" } });
  const submitButton = getByRole("button", { name: "Shorten" });
  expect(submitButton).not.toHaveAttribute("disabled");
});

test("Should show the interstitial when the service call is initiated", async () => {
  const { baseElement, getByLabelText, getByRole } = render(
    <Form setData={jest.fn()} />,
    { wrapper: MemoryRouter }
  );
  const urlField = getByLabelText(/URL:/i);
  fireEvent.change(urlField, { target: { value: "https://reactjs.org/" } });
  const submitButton = getByRole("button", { name: "Shorten" });
  act(() => {
    fireEvent.click(submitButton);
  });
  await waitFor(() => expect(baseElement).toMatchSnapshot());
});

test("Should show the error page when the service call fails", async () => {
  const { baseElement, getByLabelText, getByRole } = render(
    <Form setData={jest.fn()} />,
    { wrapper: MemoryRouter }
  );
  const urlField = getByLabelText(/URL:/i);
  fireEvent.change(urlField, { target: { value: "https://reactjs.org/" } });
  const submitButton = getByRole("button", { name: "Shorten" });
  await act(async () => {
    fireEvent.click(submitButton);
  });
  expect(baseElement).toMatchSnapshot();
});

test("Should show the error page when the service call returns a bad response", async () => {
  const { baseElement, getByLabelText, getByRole } = render(
    <Form setData={jest.fn()} />,
    { wrapper: MemoryRouter }
  );
  const urlField = getByLabelText(/URL:/i);
  fireEvent.change(urlField, {
    target: { value: "https://developer.mozilla.org/en-US/" },
  });
  const submitButton = getByRole("button", { name: "Shorten" });
  await act(async () => {
    fireEvent.click(submitButton);
  });
  expect(baseElement).toMatchSnapshot();
});

test("Should invoke the setData function when the response is valid", async () => {
  const setData = jest.fn();
  const { getByLabelText, getByRole } = render(<Form setData={setData} />, {
    wrapper: MemoryRouter,
  });
  const urlField = getByLabelText(/URL:/i);
  fireEvent.change(urlField, { target: { value: "https://www.google.com" } });
  const submitButton = getByRole("button", { name: "Shorten" });
  await act(async () => {
    fireEvent.click(submitButton);
  });
  expect(setData).toHaveBeenCalledWith({
    shortCode: "abcdef",
    target: "https://www.google.com",
  });
});

test("Should clear the form when the service call returns valid", async () => {
  const setData = jest.fn();
  const { getByLabelText, getByRole } = render(<Form setData={setData} />, {
    wrapper: MemoryRouter,
  });
  const urlField = getByLabelText(/URL:/i);
  fireEvent.change(urlField, { target: { value: "https://www.google.com" } });
  expect(urlField.value).toBe("https://www.google.com");
  const submitButton = getByRole("button", { name: "Shorten" });
  await act(async () => {
    fireEvent.click(submitButton);
  });
  expect(urlField.value).toBe("");
});

test("Should render the form after successful fetch", async () => {
  const setData = jest.fn();
  const { baseElement, getByLabelText, getByRole } = render(
    <Form setData={setData} />,
    { wrapper: MemoryRouter }
  );
  const urlField = getByLabelText(/URL:/i);
  fireEvent.change(urlField, { target: { value: "https://www.google.com" } });
  expect(urlField.value).toBe("https://www.google.com");
  const submitButton = getByRole("button", { name: "Shorten" });
  await act(async () => {
    fireEvent.click(submitButton);
  });
  expect(baseElement).toMatchSnapshot();
});
