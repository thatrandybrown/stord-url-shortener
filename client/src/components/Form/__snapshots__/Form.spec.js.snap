// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the form after successful fetch 1`] = `
<body>
  <div
    id="interstitial-root"
  />
  <div>
    <p>
      Shorten your link: provide a full url and get a shortened link for sharing.
    </p>
    <form>
      <label
        for="url"
      >
        URL:
      </label>
      <input
        id="url"
        name="url"
        required=""
        type="url"
        value=""
      />
      <input
        aria-label="Shorten"
        disabled=""
        name="submit"
        type="submit"
        value="Shorten"
      />
    </form>
  </div>
</body>
`;

exports[`Should show the error page when the service call fails 1`] = `
<body>
  <div
    id="interstitial-root"
  />
  <div>
    <section
      class="error-section"
    >
      <h2
        style="font-size: 10rem;"
      >
        Oops!
      </h2>
      <p>
        There was an error submitting your request. Please try again.
      </p>
      <a
        href="/"
      >
        Navigate Home
      </a>
    </section>
  </div>
</body>
`;

exports[`Should show the error page when the service call returns a bad response 1`] = `
<body>
  <div
    id="interstitial-root"
  />
  <div>
    <section
      class="error-section"
    >
      <h2
        style="font-size: 10rem;"
      >
        Oops!
      </h2>
      <p>
        There was an error submitting your request. Please try again.
      </p>
      <a
        href="/"
      >
        Navigate Home
      </a>
    </section>
  </div>
</body>
`;

exports[`Should show the interstitial when the service call is initiated 1`] = `
<body>
  <div
    id="interstitial-root"
  >
    <div>
      <svg
        height="64"
        version="1"
        viewBox="0 0 128 128"
        width="64"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M78.75 16.18V1.56a64.1 64.1 0 0147.7 47.7H111.8a49.98 49.98 0 00-33.07-33.08zM16.43 49.25H1.8a64.1 64.1 0 0147.7-47.7V16.2a49.98 49.98 0 00-33.07 33.07zm33.07 62.32v14.62A64.1 64.1 0 011.8 78.5h14.63a49.98 49.98 0 0033.07 33.07zm62.32-33.07h14.62a64.1 64.1 0 01-47.7 47.7v-14.63a49.98 49.98 0 0033.08-33.07z"
            fill="#b011b0"
          />
          <animatetransform
            attributeName="transform"
            dur="400ms"
            from="0 64 64"
            repeatCount="indefinite"
            to="-90 64 64"
            type="rotate"
          />
        </g>
      </svg>
    </div>
  </div>
  <div>
    <p>
      Shorten your link: provide a full url and get a shortened link for sharing.
    </p>
    <form>
      <label
        for="url"
      >
        URL:
      </label>
      <input
        id="url"
        name="url"
        required=""
        type="url"
        value="https://reactjs.org/"
      />
      <input
        aria-label="Shorten"
        name="submit"
        type="submit"
        value="Shorten"
      />
    </form>
  </div>
</body>
`;

exports[`renders instructions 1`] = `
<div>
  <p>
    Shorten your link: provide a full url and get a shortened link for sharing.
  </p>
  <form>
    <label
      for="url"
    >
      URL:
    </label>
    <input
      id="url"
      name="url"
      required=""
      type="url"
      value=""
    />
    <input
      aria-label="Shorten"
      disabled=""
      name="submit"
      type="submit"
      value="Shorten"
    />
  </form>
</div>
`;
